jmxtrans-core/src/main/java/com/googlecode/jmxtrans/cluster/ClusterConfigClient.java

```diff
@@ -0,0 +1,225 @@
+
+import java.io.File;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * The main class that menages the connection to a cluster coordination
+ * service. If the cluster connection is enabled, the configuration will
+ * be queried from the cluster coordination service.
+ */
+public class ClusterConfigClient extends Thread implements ClusterConfigChangeListener{
+    private Configuration clConfiguration;
+    private String clServiceConnectionString;
+    private CuratorFramework clClient;
+    private String workerAlias = "";
+    private String configPath = "";
+    private String heartbeatPath = "";
+    PathChildrenCache jvmRootCache;
+    PathChildrenCache workerRootCache;
+    private Map<String, ClusterJvmConfigHandler> jvmConfigs;
+
+    private static final Logger log = LoggerFactory.getLogger(ClusterConfigClient.class);
+
+    public ClusterConfigClient(){
+        jvmConfigs = new HashMap<>();
+    }
+
+    public void initialize(File clConfigFile) throws ConfigurationException{
+        this.clConfiguration = new PropertiesConfiguration(clConfigFile);
+        initialize(clConfiguration.getString("zookeper.connectionstring"),
+                clConfiguration.getString("zookeeper.workeralias"),
+                clConfiguration.getString("zookeeper.heartbeatpath"),
+                clConfiguration.getString("zookeeper.configpath"));
+    }
+
+    public void initialize(String clServiceConnectionString, String workerAlias, String heartbeatPath, String configPath) throws ConfigurationException{
+        if(null == clServiceConnectionString || "".equals(clServiceConnectionString)){
+            throw new ConfigurationException("None or empty connectionString!");
+        }
+        if(null == workerAlias || "".equals(workerAlias)){
+            throw new ConfigurationException("The worker alias cannot be null or empty! Maybe 'cluster.alias' " +
+                    "filed is missing from the configuration");
+        }
+        this.clServiceConnectionString = clServiceConnectionString;
+        this.workerAlias = workerAlias;
+        this.heartbeatPath = heartbeatPath;
+        this.configPath = configPath;
+        this.setName("Cluster-" + this.workerAlias);
+    }
+
+
+    @Override
+    public void run() {
+        try{
+            startClusterConnection();
+            startHeartBeating();
+            queryConfigs();
+
+            jvmRootCache = new PathChildrenCache(clClient, this.configPath, false);
+            workerRootCache = new PathChildrenCache(clClient, this.heartbeatPath, false);
+
+            jvmRootCache.getListenable().addListener(new JvmConfigPathChangeListener());
+            workerRootCache.getListenable().addListener(new WorkerPathChangeListener());
+            jvmRootCache.start();
+            workerRootCache.start();
+
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+
+
+        while(!this.isInterrupted()){
```

Comment:
This while loop look suspicious. You are keeping a thread alive, but not doing any actual work. It looks fishy...